File: 2024/src/day01.py
#!/usr/bin/env python3

"""
List and Counter.
"""

from collections import Counter

from aoc import AOCSolver, integers, list_transpose, map_list

# ------------------------------------------------------------------------------


class Day01Solver(AOCSolver):
    def __init__(self):
        super().__init__()

        self.day = 1
        self.expected_test = 11, 31
        self.expected = 1579939, 20351745

    def solve(self, test: bool = False) -> tuple:
        puzzle = self.puzzle.splitlines()

        a, b = list_transpose(map_list(integers, puzzle))
        a.sort()
        b.sort()

        counts = Counter(b)

        p1, p2 = 0, 0
        for x, y in zip(a, b):
            p1 += abs(y - x)
            p2 += x * counts[x]

        return p1, p2


# ------------------------------------------------------------------------------


if __name__ == "__main__":
    solver = Day01Solver()
    solver.run()


# ------------------------------------------------------------------------------

File: 2024/src/day01.zig
// -----------------------------------------------------------------------------

const std = @import("std");
const Allocator = std.mem.Allocator;

const aoc = @import("aoc");

pub const Result = struct {
    p1: u32,
    p2: u32,
};

pub fn solve(allocator: Allocator, data: []const u8, test_run: bool) !Result {
    _ = allocator;
    _ = data;
    _ = test_run;

    var result: Result = .{ .p1 = 0, .p2 = 0 };

    result.p1 = 0;
    result.p2 = 0;

    return result;
}

pub fn main() !void {
    var arena: std.heap.ArenaAllocator = .init(std.heap.page_allocator);
    defer arena.deinit();

    const allocator = arena.allocator();

    const solver: aoc.AOCSolver(Result) = .init(
        2024,
        1,
        allocator,
        solve,
    );

    solver.info();

    _ = try solver.run(true);
    // _ = try solver.run(false);
}

// EOF -------------------------------------------------------------------------
